lite2/rpc/client.go

```diff
@@ -0,0 +1,416 @@
+		return nil, err
+	}
+
+	// Validate res.
+	if err := res.SignedHeader.ValidateBasic(c.lc.ChainID()); err != nil {
+		return nil, err
+	}
+
+	// Update the light client if we're behind.
+	if err := c.updateLiteClientIfNeededTo(res.Height); err != nil {
+		return nil, err
+	}
+
+	// Verify commit.
+	h, err := c.lc.TrustedHeader(res.Height)
+	if err != nil {
+		return nil, errors.Wrapf(err, "TrustedHeader(%d)", res.Height)
+	}
+	if rH, tH := res.Hash(), h.Hash(); !bytes.Equal(rH, tH) {
+		return nil, errors.Errorf("header %X does not match with trusted header %X",
+			rH, tH)
+	}
+
+	return res, nil
+}
+
+// Tx calls rpcclient#Tx method and then verifies the proof if such was
+// requested.
+func (c *Client) Tx(hash []byte, prove bool) (*ctypes.ResultTx, error) {
+	res, err := c.next.Tx(hash, prove)
+	if err != nil || !prove {
+		return res, err
+	}
+
+	// Validate res.
+	if res.Height <= 0 {
+		return nil, errors.Errorf("invalid ResultTx: %v", res)
+	}
+
+	// Update the light client if we're behind.
+	if err := c.updateLiteClientIfNeededTo(res.Height); err != nil {
+		return nil, err
+	}
+
+	// Validate the proof.
+	h, err := c.lc.TrustedHeader(res.Height)
+	if err != nil {
+		return res, errors.Wrapf(err, "TrustedHeader(%d)", res.Height)
+	}
+	return res, res.Proof.Validate(h.DataHash)
+}
+
+func (c *Client) TxSearch(query string, prove bool, page, perPage int) (*ctypes.ResultTxSearch, error) {
+	return c.next.TxSearch(query, prove, page, perPage)
+}
+
+func (c *Client) Validators(height *int64) (*ctypes.ResultValidators, error) {
+	return c.next.Validators(height)
+}
+
+func (c *Client) BroadcastEvidence(ev types.Evidence) (*ctypes.ResultBroadcastEvidence, error) {
+	return c.next.BroadcastEvidence(ev)
+}
+
+func (c *Client) Subscribe(ctx context.Context, subscriber, query string,
+	outCapacity ...int) (out <-chan ctypes.ResultEvent, err error) {
+	return c.next.Subscribe(ctx, subscriber, query, outCapacity...)
+}
+
+func (c *Client) Unsubscribe(ctx context.Context, subscriber, query string) error {
+	return c.next.Unsubscribe(ctx, subscriber, query)
+}
+
+func (c *Client) UnsubscribeAll(ctx context.Context, subscriber string) error {
+	return c.next.UnsubscribeAll(ctx, subscriber)
+}
+
+func (c *Client) updateLiteClientIfNeededTo(height int64) error {
+	lastTrustedHeight, err := c.lc.LastTrustedHeight()
+	if err != nil {
+		return errors.Wrap(err, "LastTrustedHeight")
+	}
+	if lastTrustedHeight < height {
+		if err := c.lc.VerifyHeaderAtHeight(height, time.Now()); err != nil {
+			return errors.Wrapf(err, "VerifyHeaderAtHeight(%d)", height)
+		}
+	}
+	return nil
+}
+
+func (c *Client) RegisterOpDecoder(typ string, dec merkle.OpDecoder) {
+	c.prt.RegisterOpDecoder(typ, dec)
+}
+
+// SubscribeWS subscribes for events using the given query and remote address as
+// a subscriber, but does not verify responses (UNSAFE)!
+// TODO: verify data
+func (c *Client) SubscribeWS(ctx *rpctypes.Context, query string) (*ctypes.ResultSubscribe, error) {
```

Comment:
ditto. do we need to return the result if it's `nil`?

Reply:
yes